massa-models/src/block.rs

```diff
@@ -504,6 +505,59 @@ pub struct BlockHeader {
     pub endorsements: Vec<SecureShareEndorsement>,
 }
 
+// TODO: gh-issue #3398
+#[cfg(any(test, feature = "testing"))]
+impl BlockHeader {
+    fn assert_invariants(&self) -> Result<(), Box<dyn std::error::Error>> {
+        use crate::config::ENDORSEMENT_COUNT;
+
+        if self.slot.period == 0 {
+            if !self.parents.is_empty() {
+                return Err("Invariant broken: genesis block with parent(s)".into());
+            }
+            if !self.endorsements.is_empty() {
+                return Err("Invariant broken: genesis block with endorsement(s)".into());
+            }
+        } else {
+            if self.parents.len() != crate::config::THREAD_COUNT as usize {
+                return Err(
+                    "Invariant broken: non-genesis block with incorrect number of parents".into(),
+                );
+            }
+            if self.endorsements.len() > crate::config::ENDORSEMENT_COUNT as usize {
+                return Err("Invariant broken: endorsement count too high".into());
+            }
+
+            let parent_id = self.parents[self.slot.thread as usize];
+            for endo in self.endorsements.iter() {
+                if endo.content.endorsed_block != parent_id {
+                    return Err("Invariant broken: endorsement doesn't match parent".into());
+                }
+            }
+        }
+
+        // assert that the endorsement indexes are all unique...
+        let is_idx_set = self
+            .endorsements
+            .iter()
+            .map(|endo| {
+                endo.verify_signature()?;
+                // ..and that they are within range
+                if endo.content.index >= ENDORSEMENT_COUNT {
+                    return Err::<u32, Box<dyn std::error::Error>>(
+                        "Invariant broken: endorsement index out of bounds".into(),
+                    );
+                }
+                Ok(endo.content.index)
+            })
+            .try_collect::<std::collections::HashSet<_>>()?
+            .len()
+            == self.endorsements.len();
```

Comment:
There has to be a better way, such as an `try_into_set` or a `try_map`, but nothing jumped out at me at a quick glance